core/src/main/scala/cats/Eval.scala

```diff
@@ -178,6 +178,13 @@ object Always {
 
 object Eval extends EvalInstances {
 
+  /**
+   * A Leaf does not depend on any other Eval
+   * so calling .value does not trigger
+   * any flatMaps or defers
+   */
+  sealed abstract class Leaf[A] extends Eval[A]
```

Comment:
with this, we can use `x: Leaf[A]` below to make it clear calling `.value` is safe. This helps the reader review the `def evaluate` method and see that it is safe (we only call .value on Leafs).

it doesn't impact the performance (hopefully scala removes the type check when it can statically prove that it would always pass, and since this was in the last position of the match that is the case currently).